### PR TITLE
feat: add disk chopper unit validator

### LINKED ISSUE #3: Add check for correct units in NXdisk_chopper
I have seen some files with missing frequency units (or units '' instead of 'Hz').

### PR DESCRIPTION
**Pull request by jokasimr**
Fixes #3

### CODE REVIEW

File: src/chexus/validators.py [hunk 1]
```diff
@@ -205,6 +205,28 @@ def validate(self, node: Dataset | Group) -> Violation | None:
...
+        return (
+            isinstance(node, Group)
+            and node.attrs.get('NX_class') == 'NXdisk_chopper'
+            and 'rotation_speed' in node.children
+        )
+
+    def validate(self, node: Dataset | Group) -> Violation | None:
+        if 'units' in node.children.get('rotation_speed').attrs:
+            unit = node.children.get('rotation_speed').attrs.get('units')
+            if unit.endswith('Hz') or unit.endswith('s') and '/' in unit:
```
Comment (jl-wynen) [inline src/chexus/validators.py:225]: These checks are too loose. E.g., these two would erroneously pass: `s*Hz`, `m/s`

**Review by jl-wynen — CHANGES_REQUESTED** [review]

File: src/chexus/validators.py (hunk 1 above)
Comment (jokasimr) [reply src/chexus/validators.py:225]: I think it make sense to bring in [pint](https://pint.readthedocs.io/en/stable/getting/index.html) as a dependency. It's likely we will want to do similar checks in other places.

Comment (jl-wynen) [reply src/chexus/validators.py:225]: Since this is a pure analysis tool, i.e., it won't affect any actual loaders, I'm for being strict.

Maybe use a regex like this: `\w*(1\w*/\w*s|[umkM]?(Hz|hz|HZ))\w*`. Though realistically, the SI prefix for Hz is unlikely to be used. The hoppers we have cosily spin at a few Hz.

Comment (jokasimr) [reply src/chexus/validators.py]: I think the regex you sent would probably work very well in this case. But I still lean towards using a library for this since it's more likely to handle surprising edge cases correctly and it seems likely we will want to do similar unit checks in other validators anyway.

Comment (jl-wynen) [reply src/chexus/validators.py]: While I in principle agree, we need to be careful with using a different parser than Scipp's because the same edge cases that you are worried about can lead to differences in the parsers. So I'd rather we use Scipp for this. Or have a clear decision against using Scipp, then you can go ahead with Pint.

@SimonHeybrock ?

Comment (jokasimr) [reply src/chexus/validators.py]: I'm all for using scipp instead of pint here. I opted for pint in this case because I got the impression that it was a priority to keep the dependencies  lightweight.

Comment (SimonHeybrock) [reply src/chexus/validators.py]: I have suggested elsewhere that we should use Scipp to parse units, but make the tool still work without. That is, either dynamically skip the checks, or add a `--use-scipp` command line option to enable additional checks.

Comment (jokasimr) [reply src/chexus/validators.py]: I'd say make checking units the default and if `scipp` can't be found then either
  1) skip the tests that require it and inform the user what happened and what they need to do to run the full test suite, or
  2) fail and inform the user they need to use a command line option like `--no-scipp` to run the tool without unit checking.

Making it opt-in I think there's risk the option will not be discovered.

Comment (SimonHeybrock) [reply src/chexus/validators.py]: Option 2 sounds good, I think?

**Review by SimonHeybrock — COMMENTED** [review]
Now that we can use Scipp, we can add a validator that units can be parsed?

**Comment by jokasimr** [comment]
> Now that we can use Scipp, we can add a validator that units can be parsed?

For chopper disks or for all nodes?
For chopper disks there is one already.

File: src/chexus/__main__.py [hunk 2]
```diff
@@ -20,9 +20,28 @@ def main():
     parser.add_argument(
         '--checksums', action='store_true', help='Compute and print checksums'
     )
+    parser.add_argument(
+        '--no-scipp', action='store_true', help='Skip the validators that require scipp'
```
Comment (SimonHeybrock) [inline src/chexus/__main__.py:24]: You missed the obvious pun of using `--skip-scipp`.

Just kidding, please don't do this ;)

File: src/chexus/__main__.py [hunk 3]
```diff
@@ -20,9 +20,28 @@ def main():
...
     args = parser.parse_args()
     path = args.path
+    no_scipp = args.no_scipp
+
+    try:
+        import scipp  # noqa: F401
+    except ModuleNotFoundError:
+        if not no_scipp:
+            print(
+                'Warning: Scipp was not found. The Nexus file validation was not run.\n'
```
Comment (SimonHeybrock) [inline src/chexus/__main__.py:36]:
```suggestion
                'Error: Scipp was not found. The Nexus file validation was not run.\n'
```

File: src/chexus/__main__.py [hunk 4]
```diff
@@ -20,9 +20,28 @@ def main():
...
+                'To run the full test suite you need to install scipp'
+                ' using `pip install scipp` or `conda install -c scipp scipp`.'
+                ' This is recommended.\n'
+                'To run only the tests that don\'t require scipp and silence this'
+                ' warning add the flag `--no-scipp`.'
```
Comment (SimonHeybrock) [inline src/chexus/__main__.py:41]: Since it exists with an error, it is not about silencing the warning, is it?

File: src/chexus/__main__.py (hunk 2 above)
Comment (jokasimr) [reply src/chexus/__main__.py:24]: I was sorely tempted...

File: src/chexus/validate.py [hunk 5]
```diff
@@ -63,8 +63,13 @@ def validate(group: Group, validators: list[Validator]) -> dict[type, Validation
...
     for node in tree.values():
-        for validation in results.values():
-            validation.apply(node)
+        try:
+            for validation in results.values():
+                validation.apply(node)
+        except ModuleNotFoundError as e:
+            if e.name != 'scipp':
+                raise e
+
```
Comment (SimonHeybrock) [inline src/chexus/validate.py:72]: Why do we need this?

Comment (jokasimr) [reply src/chexus/validate.py:72]: Because if we don't have it the `ModuleNotFoundError` will be raised when chexus is called will `--no-scipp`. We don't want that, we just want to continue with the next validator.

Comment (SimonHeybrock) [reply src/chexus/validate.py]: Can we just remove all validators that depend on scipp, instead of going via try...except?

Comment (jokasimr) [reply src/chexus/validate.py]: We could but then we'd have to remember to register some validators as "depending on scipp", I just found the try... except better because it avoids that. The disadvantage I see with the try except version is that we have to import `scipp` in the validator where it's needed.

An advantage of doing it this way is that in the report we get an entry like
``name_of_violation_depending_on_scipp ?/5 (requires scipp)``
if there is a validator that is applicable 5 times but couldn't be run because scipp is missing.

We could of course implement that feature in some other way as well, but if we're adding extra logic anyway I don't see why the try except is worse than anything else.

File: src/chexus/validate.py [hunk 6]
```diff
@@ -64,7 +67,13 @@ def validate(group: Group, validators: list[Validator]) -> dict[type, Validation
...
     for node in tree.values():
         for validation in results.values():
-            validation.apply(node)
+            try:
+                validation.apply(node)
+            except ModuleNotFoundError as e:
+                if e.name != 'scipp':
+                    raise e
+                validation.skipped = True
+
```
Comment (SimonHeybrock) [inline src/chexus/validate.py:76]: Instead of this, maybe add an argument to `base_validators` that deselecteds scipp-based validators if the flag is set.

Comment (jokasimr) [reply src/chexus/validate.py:76]: I thought about that but decided against it for the reasons mentioned above. I'm open to doing it that way but I'm curious to know why you think that's better?

Comment (SimonHeybrock) [reply src/chexus/validate.py:76]: Many people (maybe fewer in Python than in other languages) feel that doing "regular" control flow via exceptions is an anti-pattern (it should be "exceptions", as the name implies). In this case you can well encounter this try-except 10s or 100s of thousands of times for any given file, which is also not so nice.

File: src/chexus/validate.py (hunk 5 above)
Comment (SimonHeybrock) [reply src/chexus/validate.py]: You can still list something like `Skipped validators (via --no-scipp): validator1, validator2, ...`. Knowing how many nodes it *would* have applied to does not add too much benefit?

Comment (jokasimr) [reply src/chexus/validate.py]: > Knowing how many nodes it would have applied to does not add too much benefit?

I think that particularly adds benefit because if there are many nodes where it applies then the user can estimate the value of rerunning with scipp installed. And vice versa if it applies zero times the user can be confident in the results even if `--no-scipp` was set.

File: src/chexus/validate.py (hunk 6 above)
Comment (jokasimr) [reply src/chexus/validate.py:76]: I agree that it's an anti pattern if there's another canonical way to do the control flow. (The classical example would be trying to get a value from a dictionary that maybe doesn't exist, and using try/except instead of `get`). However, for control flow regarding non-avaliable imports I'd say try except doesn't really have an equivalent canonical way.

Regarding performance I think that's unlikely to be an issue. My understanding is that try-except only has performance implications if the exception is triggered. That only happens when the scipp based validators are applied to nodes where they are applicable and that's unlikely to be a large fraction of the total.

File: src/chexus/validate.py (hunk 5 above)
Comment (SimonHeybrock) [reply src/chexus/validate.py]: I think you are overthinking this, most users will be happy to install (or already have installed) scipp.

**Review by SimonHeybrock — APPROVED** [review]

File: src/chexus/validate.py (hunk 5 above)
Comment (jokasimr) [reply src/chexus/validate.py]: Yeah you are probably right :sweat_smile:

File: src/chexus/validate.py (hunk 6 above)
Comment (SimonHeybrock) [reply src/chexus/validate.py:76]: Note that your current approach will also break if someone writes a validator that imports in one of the other methods.

File: src/chexus/validate.py [hunk 7]
```diff
@@ -56,6 +57,8 @@ def format_details(self) -> str:
         return details
 
     def format_summary(self) -> str:
+        if self.skipped:
+            return f"{self.validator.name}: ?/{self.checks} (requires scipp)\n"
```
Comment (jl-wynen) [inline src/chexus/validate.py:61]: You can't technically tell whether this skips because scipp is unavailable of for a different reason. Minor comment, though, I leave it up to you to decide what to do.

File: src/chexus/validate.py (hunk 6 above)
Comment (jl-wynen) [reply src/chexus/validate.py:76]: I agree that using excepts to handle import failures is idiomatic. But then it should be tightly placed around the import. Here, you can't tell which import raised the exception.
So how about checking whether scipp is available and whether the command line argument is set and then setting a config option somewhere? (I'm not familiar enough with the code to say where that would go.) And then the validators that use scipp would check that flag and not even run if the check is disabled.

File: tests/validators_test.py
Comment (jl-wynen) [inline tests/validators_test.py:1]: Can you add a CI run without scipp to check that the `--no-scipp` argument is handled properly?

File: src/chexus/validate.py (hunk 7 above)
Comment (jokasimr) [reply src/chexus/validate.py:61]: I considered replacing the `self.skipped` boolean with an optional `self.skipped_reason` string. Then the function body would be replaced by
```python
if self.skipped_reason:
    return f"{self.validator.name}: ?/{self.checks} ({self.skipped_reason})\n"
```
with the advantage that we have to specify the reason for skipping in the same place where we mark the validator as skipped.

However, since we only skip validators for a single reason right now (scipp being unavaliable) that felt a bit overkill.

**Comment by jokasimr** [comment]
Me and @jl-wynen talked a bit about this and since both he and @SimonHeybrock didn't really like the current approach we decided to change it.
I think one of the root questions was whether or not the validators should know under what conditions they are skipped. I saw it as a good thing that the validators did not have to know or implement their own skipping logic, but I see how that might make the behavior harder to test.

File: src/chexus/validate.py (hunk 7 above)
Comment (SimonHeybrock) [reply src/chexus/validate.py:61]: I think the (length of the) whole discussion shows that the try/except approach and the `skipped` flag are causing more trouble than they have benefit. I suggest you simply select validators based on what libraries they require (as suggested earlier), then the whole discussion it moot.

**Comment by SimonHeybrock** [comment]
> I think one of the root questions was whether or not the validators should know under what conditions they are skipped. I saw it as a good thing that the validators did not have to know or implement their own skipping logic, but I see now how that might make the behavior harder to test.

If this worries you, add a `requirements` (class) attr or method to each validator, which they can list required libs (or other things). Then the logic that builds a list of validators can check whether those features are available and decide whether to include the validators.

**Comment by jokasimr** [comment]
> If this worries you, add a requirements (class) attr or method to each validator, which they can list required libs (or other things). Then the logic that builds a list of validators can check whether those features are available and decide whether to include the validators.

I don't understand, I think we're talking past each other here?

It seems to me that what you are suggesting is that the validator runner / list builder, i.e. not the validator itself, determines if it should be skipped or not. But that is what is currently implemented, except that the 'list of requirements' is automatically inferred at runtime by catching the import failure.

**Comment by SimonHeybrock** [comment]
> > If this worries you, add a requirements (class) attr or method to each validator, which they can list required libs (or other things). Then the logic that builds a list of validators can check whether those features are available and decide whether to include the validators.
> 
> I don't understand, I think we're talking past each other here?
> 
> It seems to me that what you are suggesting is that the validator runner / list builder, i.e. not the validator itself, determines if it should be skipped or not. But that is what is currently implemented, except that the 'list of requirements' is automatically inferred at runtime by catching the import failure.

I was suggesting to do this at the time when we build the list of validators. Current it is done for every dataset, when the validators are *called*.

But I am fine with just building those lists "by hand", logic for having validators indicate what they require is probably overkill. Keep it simple for now. I think I'd simply modify https://github.com/scipp/chexus/blob/245fd376d5e2f741a6fb2585e32e21a1b1a47fa1/src/chexus/validators.py#L261-L275 with such a check.

File: src/chexus/validators.py [hunk 8]
```diff
@@ -271,5 +299,9 @@ def base_validators():
         NX_class_is_legacy(),
         transformation_depends_on_missing(),
         transformation_offset_units_missing(),
-        units_invalid(),
```
Comment (SimonHeybrock) [inline src/chexus/validators.py:274]: Why was this removed?

Comment (jokasimr) [reply src/chexus/validators.py:274]: No reason, I've added it back.

**Review by SimonHeybrock — APPROVED** [review]
